sdl_android/src/main/java/com/smartdevicelink/api/screen/SoftButtonManager.java

```diff
@@ -0,0 +1,549 @@
+            for (SoftButtonObject softButtonObject : softButtonObjects) {
+                SoftButtonState initialState = null;
+                if (softButtonObject != null) {
+                    initialState = softButtonObject.getCurrentState();
+                }
+                if (initialState != null && softButtonObject.getStates() != null) {
+                    for (SoftButtonState softButtonState : softButtonObject.getStates()) {
+                        if (softButtonState != null && softButtonState.getName() != null && softButtonState.getArtwork() != null && !fileManager.hasUploadedFile(softButtonState.getArtwork())) {
+                            if (softButtonState.getName().equals(initialState.getName())) {
+                                initialStatesToBeUploaded.add(softButtonObject.getCurrentState().getArtwork());
+                            } else{
+                                otherStatesToBeUploaded.add(softButtonState.getArtwork());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+
+        // Upload initial state images
+        if (initialStatesToBeUploaded.size() > 0) {
+            Log.v(TAG, "Uploading soft button initial state artworks");
+            fileManager.uploadArtworks(initialStatesToBeUploaded, new MultipleFileCompletionListener() {
+                @Override
+                public void onComplete(Map<String, String> errors) {
+                    if (errors != null && errors.size() > 0) {
+                        Log.e(TAG, "Error uploading soft button artworks");
+                    }
+                    Log.d(TAG, "Soft button initial artworks uploaded");
+                    update(cachedListener);
+                }
+            });
+        }
+
+
+        // Upload other state images
+        if (otherStatesToBeUploaded.size() > 0) {
+            Log.v(TAG, "Uploading soft button other state artworks");
+            fileManager.uploadArtworks(otherStatesToBeUploaded, new MultipleFileCompletionListener() {
+                @Override
+                public void onComplete(Map<String, String> errors) {
+                    if (errors != null && errors.size() > 0) {
+                        Log.e(TAG, "Error uploading soft button artworks");
+                    }
+                    Log.d(TAG, "Soft button other state artworks uploaded");
+                    // In case our soft button states have changed in the meantime
+                    update(cachedListener);
+                }
+            });
+        }
+
+        // This is necessary because there may be no images needed to be uploaded
+        update(cachedListener);
+    }
+
+    /**
+     * Update the SoftButtonManger by sending a new Show RPC to reflect the changes
+     * @param listener a CompletionListener
+     */
+    protected void update(CompletionListener listener) {
+        cachedListener = listener;
```

Comment:
This will get set to `null` sometimes, is that acceptable? For instance, out of the SoftButonObject.

Reply:
I believe this should also only be set before the next two `return` statements.

Reply:
I don't see why this could be an issue. When the SoftButtonObject calls update(null), that shouldn't trigger the callback (same behavior in iOS)